File: python/src/x402/facilitatorclient/facilitatorclient.py
import requests

from ..types import PaymentRequirements, PaymentPayload

# DefaultFacilitatorURL is the default URL for the x402 facilitator service
# DEFAULT_FACILITATOR_URL = "https://x402.org/facilitator"
DEFAULT_FACILITATOR_URL = "http://localhost:3002"

class FacilitatorClient:
    def __init__(self, url=None):
        self.url = url or DEFAULT_FACILITATOR_URL

    def verify(self, payment_payload: PaymentPayload, payment_requirements: PaymentRequirements):
        """
        Verifies a payment with the facilitator

        Parameters
        ----------
        payment_payload: PaymentPayload
            The payment payload to verify
        payment_requirements: PaymentRequirements
            The payment requirements to verify against

        Returns
        -------
        dict
            The response from the facilitator

        Raises
        ------
        Exception
            If the request to the facilitator fails
        """
        req_body = {
            "x402Version": 1,
            "paymentPayload": payment_payload.to_dict() if isinstance(payment_payload, PaymentPayload) else payment_payload,
            "paymentRequirements": payment_requirements.to_dict() if isinstance(payment_requirements, PaymentRequirements) else payment_requirements
        }

        response = requests.post(f"{self.url}/verify", json=req_body)
        if response.status_code != 200:
            raise Exception(f"Failed to verify payment: {response.status_code} - {response.text}")

        return response.json()

    def settle(self, payment_payload: PaymentPayload, payment_requirements: PaymentRequirements):
        """
        Settles a payment with the facilitator

        Parameters
        ----------
        payment_payload: PaymentPayload
            The payment payload to settle
        payment_requirements: PaymentRequirements
            The payment requirements to settle against

        Returns
        -------
        dict
            The response from the facilitator

        Raises
        ------
        Exception
            If the request to the facilitator fails
        """
        req_body = {
            "x402Version": 1,
            "paymentPayload": payment_payload,
            "paymentRequirements": payment_requirements.to_dict() if isinstance(payment_requirements, PaymentRequirements) else payment_requirements
        }

        response = requests.post(f"{self.url}/settle", json=req_body)
        if response.status_code != 200:
            raise Exception(f"Failed to settle payment: {response.status_code} - {response.text}")

        return response.json()

File: php/x402/src/Http/FacilitatorClient.php
<?php

declare(strict_types=1);

namespace X402\Http;

use GuzzleHttp\Client;
use GuzzleHttp\Exception\GuzzleException;
use X402\Exception\SettlementFailedException;
use X402\Exception\VerificationFailedException;
use X402\Exception\X402Exception;
use X402\Payload\PaymentPayload;
use X402\Payload\PaymentRequirements;
use X402\Payload\SettleResponse;
use X402\Payload\VerifyResponse;
use X402\X402;

final readonly class FacilitatorClient
{
    public const string DEFAULT_URL = 'https://x402.org/facilitator';
    private string $baseUrl;

    public function __construct(
        private Client $client = new Client(['timeout' => 30]),
        string $baseUrl = self::DEFAULT_URL
    ) {
        $this->baseUrl = rtrim($baseUrl, '/');
    }

    public function verify(PaymentPayload $payment, PaymentRequirements $requirements): VerifyResponse
    {
        $data = $this->postJson('/verify', [
            'x402Version' => X402::VERSION,
            'paymentPayload' => $payment->toArray(),
            'paymentRequirements' => $requirements->toArray(),
        ]);

        $response = VerifyResponse::fromArray($data);

        if (!$response->isValid) {
            throw new VerificationFailedException(
                $response->invalidReason ?? 'Unknown error',
                $response->payer,
            );
        }

        return $response;
    }

    public function settle(PaymentPayload $payment, PaymentRequirements $requirements): SettleResponse
    {
        $data = $this->postJson('/settle', [
            'x402Version' => X402::VERSION,
            'paymentPayload' => $payment->toArray(),
            'paymentRequirements' => $requirements->toArray(),
        ]);

        $response = SettleResponse::fromArray($data);

        if (!$response->success) {
            throw new SettlementFailedException($response->errorReason ?? 'Unknown error');
        }

        return $response;
    }

    /**
     * @return array<string, mixed>
     */
    public function supported(): array
    {
        return $this->getJson('/supported');
    }

    /**
     * @param array<string, mixed> $body
     * @return array<string, mixed>
     */
    private function postJson(string $path, array $body): array
    {
        try {
            $response = $this->client->post($this->baseUrl . $path, [
                'json' => $body,
            ]);

            $decoded = json_decode($response->getBody()->getContents(), true);
            if (!is_array($decoded)) {
                throw new X402Exception('Invalid JSON response');
            }
            /** @var array<string, mixed> $decoded */
            return $decoded;
        } catch (GuzzleException $e) {
            throw new X402Exception("HTTP request failed: {$e->getMessage()}", 0, $e);
        }
    }

    /**
     * @return array<string, mixed>
     */
    private function getJson(string $path): array
    {
        try {
            $response = $this->client->get($this->baseUrl . $path);

            $decoded = json_decode($response->getBody()->getContents(), true);
            if (!is_array($decoded)) {
                throw new X402Exception('Invalid JSON response');
            }
            /** @var array<string, mixed> $decoded */
            return $decoded;
        } catch (GuzzleException $e) {
            throw new X402Exception("HTTP request failed: {$e->getMessage()}", 0, $e);
        }
    }
}
